core/runtime/common/module_repository_impl.cpp

```diff
@@ -93,30 +57,71 @@ namespace kagome::runtime {
           return code.as_failure();
         }
         OUTCOME_TRY(new_module, module_factory_->make(code.value()));
-        module = std::move(new_module);
-        BOOST_VERIFY(modules_.put(state, module));
-      } else {
-        module = opt_module.value();
+        BOOST_VERIFY(
+            runtime_instances_pool_.put_module(state, std::move(new_module)));
       }
     }
+
+    // Try acquire instance (instantiate if needed)
+    OUTCOME_TRY(runtime_instance, runtime_instances_pool_.try_acquire(state));
     KAGOME_PROFILE_END(module_retrieval)
 
-    KAGOME_PROFILE_START(module_instantiation)
-    std::shared_ptr<ModuleInstance> shared_instance;
-    {
-      std::lock_guard guard{instances_mutex_};
-      OUTCOME_TRY(instance, module->instantiate());
-      shared_instance = std::move(instance);
+    return runtime_instance;
+  }
+
+  outcome::result<std::shared_ptr<ModuleInstance>>
+  RuntimeInstancesPool::try_acquire(
+      const RuntimeInstancesPool::RootHash &state) {
+    std::scoped_lock guard{mt_};
+    auto tid = soralog::util::getThreadNumber();
+    auto &pool = pools_[state];
+
+    // if used instance requested, just return it - don't add to TLS
```

Comment:
```suggestion
    // if an instance already in use requested, just return it - don't add to TLS
```
I believe that's slightly clearer